src/topology/uniform_space/basic.lean

```diff
@@ -2,38 +2,121 @@
+defined as `{ p : X × X | ∃ z, (p.1, z) ∈ V ∧ (z, p.2) ∈ W }`.
+In the metric space case, if `V = { p | dist p.1 p.2 ≤ r }` and `W = { p | dist p.1 p.2 ≤ r' }`
+then the triangle inequality, as reformulated above, says `V ○ W` is contained in
+`{p | dist p.1 p.2 ≤ r + r'}` which is the entourage associated to the radius `r + r'`.
+In general we have `mem_ball_comp (h : y ∈ ball x V) (h' : z ∈ ball y W) : z ∈ ball x (V ○ W)`.
+Note that this discussion does not depend on any axiom imposed on the uniformity filter,
+it is simply captured by the definition of composition.
+
+The uniform space axioms ask the filter `𝓤 X` to satisfy:
+* every `V ∈ 𝓤 X` contains the diagonal `id_rel = { p | p.1 = p.2}`. This abstracts the fact
+  that `dist x x ≤ r` for every non-negative radius `r` in the metric space case, and also that
+  `x - x` belongs to every neighborhood of zero in the toplogical group case.
+* `V ∈ 𝓤 X → prod.swap '' V ∈ 𝓤 X`. This is tightly related the fact that `dist x y = dist y x`
+  in a metric space, and to continuity of negation in the topological group case.
+* `∀ V ∈ 𝓤 X, ∃ W ∈ 𝓤 X, W ○ W ⊆ V`. In the metric space case, it corresponds
+  to the possibility of cuting a radius in half and the triangle inequality.
+  In the topological group case, it comes from continuity of addition at `(0, 0)`.
+
+These three axioms are stated more abstractly in the definition below, in terms of
+filters operations, without directly manipulating entourages.
+
+## Main definitions
+
+* `uniform_space X` is a uniform space structure on a type `X`
+* `uniform_continuous f` is a predicate saying a function `f : α → β` between uniform spaces
+  is uniformly continuous : `∀ r ∈ 𝓤 β, ∀ᶠ (x : α × α) in 𝓤 α, (f x.1, f x.2) ∈ r`
+
+In this file we also define a complete lattice structure on the type `uniform_space X`
+of uniform structures on `X`, as well as the uniform structure pull back operation
+`uniform_space.comap` coming from filters pull-back.
+Like distance functions, uniform structure cannot be pushed foward in general.
+
+## Notations
+
+Localized in `uniformity`, we have the notation `𝓤 X` for the uniformity on a uniform space `X`,
+and `○` for composition of relations, seen as terms with type `set (X × X)`.
+
+## Implementation notes
+
+There is already a theory of relations in `data/rel.lean` where the main definition is
+`def rel (α β : Type*) := α → β → Prop`.
+The relations used in the current file involve only one type, but this is not the reason why
+we don't reuse `data/rel.lean`. We use `set (α × α)`
+instead `rel α α` because we really need sets to use the filter library, and elements
+of filters on `α × α` have type `set (α × α)`.
+
+The structure `uniform_space X` bundles a uniform structure on `X`, a topology on `X` and
+an assumption saying those are compatible. This deos not mathematically reasonable at first,
+but is an instance of the forgetful inheritance pattern, see the explanation in
```

Comment:
```suggestion
an assumption saying those are compatible. This may not seem mathematically reasonable at first,
but is in fact an instance of the forgetful inheritance pattern. See the explanation in
```